src/Utility/Language.php

```diff
@@ -13,33 +13,59 @@ class Language {
 	 *
 	 * @return array
 	 */
-	public static function parseLanguageList($languageList = null, $forceLowerCase = true) {
+	public static function parseLanguageList($languageList = null, $forceAllLowerCase = true, $keepDuplicates = true) {
 		if ($languageList === null) {
 			if (empty(env('HTTP_ACCEPT_LANGUAGE'))) {
 				return [];
 			}
 			$languageList = env('HTTP_ACCEPT_LANGUAGE');
 		}
 		$languages = [];
+		$languagesRanks = [];
 		$languageRanges = explode(',', trim($languageList));
 		foreach ($languageRanges as $languageRange) {
 			$pattern = '/(\*|[a-zA-Z0-9]{1,8}(?:-[a-zA-Z0-9]{1,8})*)(?:\s*;\s*q\s*=\s*(0(?:\.\d{0,3})|1(?:\.0{0,3})))?/';
 			if (preg_match($pattern, trim($languageRange), $match)) {
 				if (!isset($match[2])) {
-					$match[2] = '1.0';
+					$rank = '1.0';
 				} else {
-					$match[2] = (string)(float)($match[2]);
+					$rank = (string)(float)($match[2]);
 				}
-				if (!isset($languages[$match[2]])) {
-					if ($match[2] === '1') {
-						$match[2] = '1.0';
+				if (!isset($languages[$rank])) {
+					if ($rank === '1') {
+						$rank = '1.0';
 					}
-					$languages[$match[2]] = [];
+					$languages[$rank] = [];
 				}
-				if ($forceLowerCase) {
-					$languages[$match[2]][] = strtolower($match[1]);
+
+				$language = $match[1];
+				if ($forceAllLowerCase) {
+					$language = strtolower($language);
 				} else {
-					$languages[$match[2]][] = $match[1];
+					$language = substr_replace($language, strtolower(substr($language, 0, 2)), 0, 2);
```

Comment:
Well, browser locales are defined as "xx-yy" always, as you can see from the browser data coming in.
So that was the assumption this code worked on.
We could normalize in one direction. Internally we usually use "xx_YY", so I am fine with the underscore.